### PR TITLE
JDK-8315279: Factor 'basic_plus_adr' out of PhaseMacroExpand and delete make_load/store

### PR DESCRIPTION
**Pull request by JohnTortugo**
I believe the factory methods for AddPNode should be in the AddPNode class. The make_load / make_store methods in PhaseMacroExpand can be refactored to instead just use the "make" methods from Load/Store classes.

Tested with tier1-3.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315279](https://bugs.openjdk.org/browse/JDK-8315279): Factor 'basic_plus_adr' out of PhaseMacroExpand and delete make_load/store (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [5d988b01](https://git.openjdk.org/jdk/pull/15480/files/5d988b011cd1df5226fa43b826e36fd801fda1d0)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to [5d988b01](https://git.openjdk.org/jdk/pull/15480/files/5d988b011cd1df5226fa43b826e36fd801fda1d0)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15480/head:pull/15480` \
`$ git checkout pull/15480`

Update a local copy of the PR: \
`$ git checkout pull/15480` \
`$ git pull https://git.openjdk.org/jdk.git pull/15480/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15480`

View PR using the GUI difftool: \
`$ git pr show -t 15480`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15480.diff">https://git.openjdk.org/jdk/pull/15480.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15480#issuecomment-1698427938)

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Good cleanup.
My only suggestion is to keep `AddPNode::make()` and `PhaseValues::makecon()` in header files .hpp - they are small and no need to put to .cpp files.

**Review by vnkozlov — APPROVED** [review]
Good. You need second review.

**Review by chhagedorn — APPROVED** [review]
Otherwise, the clean-up looks good.

File: src/hotspot/share/opto/addnode.hpp
Comment (JohnTortugo) [reply src/hotspot/share/opto/addnode.hpp:177]: That's a good idea. Thanks!

File: src/hotspot/share/opto/macro.cpp
Comment (JohnTortugo) [reply src/hotspot/share/opto/macro.cpp:1683]: Yeah, I actually tried that but the reuse of make_load/store wasn't justifiable without a much bigger refactor. I'll leave that refactor for another PR.

File: src/hotspot/share/opto/addnode.hpp
Comment (iwanowww) [inline src/hotspot/share/opto/addnode.hpp:181]: The overloads do look confusing now. When seeing a call to a factory method, I expect to see all inputs to be separately initialized. With `basic_plus_adr` the intention was clearer.

I'm not against the refactoring, but it feels like it moves the code in the wrong direction. The root problem is that we can't use `GraphKit `during macro expansion, so we duplicated functionality there over time. (`PhaseIdealLoop` is another victim.)

A better way would be to use `GraphKit` directly. And when we discussed it last time [1], Roland mentioned it was already implemented in Valhalla repo. 

[1] https://mail.openjdk.org/pipermail/hotspot-compiler-dev/2020-August/039612.html

Comment (iwanowww) [reply src/hotspot/share/opto/addnode.hpp:181]:
```
static Node* make(PhaseIterGVN& igvn, Node* base, Node* ptr, Node* offset) {
    return igvn.register_new_node_with_optimizer(new AddPNode(base, ptr, offset));
  }
```
vs
```
  static Node* make(PhaseIterGVN& igvn, Node* base, int offset) {
    return (offset == 0) ? base : make(igvn, base, base, (Node*) igvn.makecon(offset));
  }
```

IMO latter case adds more confusion than improves readability. Why not  unconditionally call `AddPNode::make(base, base, offset)` instead?

Comment (JohnTortugo) [reply src/hotspot/share/opto/addnode.hpp:181]: Thanks for clarifying @iwanowww . I think I see your point now. My original intent was to just move these methods out of PhaseMacroExpand and not much else. 

I'm going to do some more refactoring and patch all users of these make methods to just use this single method: `static Node* make(PhaseIterGVN& igvn, Node* base, Node* ptr, Node* offset)`. What do you think?

Comment (iwanowww) [reply src/hotspot/share/opto/addnode.hpp:181]: Sounds good. 

But in the future I'd like to see `PhaseMacroExpand` and `PhaseIdealLoop` migrated to `GraphKit` instead.

Comment (JohnTortugo) [reply src/hotspot/share/opto/addnode.hpp:181]: Actually, I think it's silly to create a method to just do a "return ... new AddP":
```
static Node* make(PhaseIterGVN& igvn, Node* base, Node* ptr, Node* offset) {
    return igvn.register_new_node_with_optimizer(new AddPNode(base, ptr, offset));
}
```
I can just replace the uses of this `make` with a `new AddP...`.

> But in the future I'd like to see PhaseMacroExpand and PhaseIdealLoop migrated to GraphKit instead.

Roger that.

Comment (JohnTortugo) [reply src/hotspot/share/opto/addnode.hpp]: @iwanowww do you think I should just withdraw this PR and close the associated RFE?

Comment (iwanowww) [reply src/hotspot/share/opto/addnode.hpp]: It's up to you, Cesar. I find existing code good enough, but also I don't have anything against your proposal.

Comment (JohnTortugo) [reply src/hotspot/share/opto/addnode.hpp]: Thanks, Vladimir. Since the work is already done and it may improve things slightly, I'll keep the PR open.
I created this RFE for migrating PhaseMacroExpand, PhaseIdealLoop to GraphKit: https://bugs.openjdk.org/browse/JDK-8316560

**Comment by vnkozlov** [comment]
I don't like the last update. Now you may create AddP node even if offset is 0. Yes, AddPNode::Identity() will fix that but it will happen only later during IGVN transform. The only code which checks offset in BarrierSetC2::obj_allocate is hard to read.

The duplicated code was factored into `basic_plus_adr()` for that reason - to simplify code in uses.

May be we should put these changes on hold until after you look on JDK-8316560.

**Pull request closed without merging**
